substrate/frame/examples/offchain-worker/src/lib.rs

```diff
@@ -305,34 +312,93 @@ pub mod pallet {
+		}
+	}
+
+	impl<T: Config> Default for ValidateUnsignedPriceSubmission<T> {
+		fn default() -> Self {
+			Self(Default::default())
+		}
+	}
 
-		/// Validate unsigned call to this module.
-		///
-		/// By default unsigned transactions are disallowed, but implementing the validator
-		/// here we make sure that some particular calls (the ones produced by offchain worker)
-		/// are being whitelisted and marked as valid.
-		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+	impl<T: Config> ValidateUnsignedPriceSubmission<T> {
+		/// Create a new instance of the transaction extension.
+		pub fn new() -> Self {
+			Self(Default::default())
+		}
+	}
+
+	impl<T: Config + Send + Sync> TransactionExtension<Call<T>> for ValidateUnsignedPriceSubmission<T>
+	where
+		Call<T>: Dispatchable<
+			Info = frame_support::dispatch::DispatchInfo,
+			PostInfo = frame_support::dispatch::PostDispatchInfo,
+		>,
+		<Call<T> as Dispatchable>::RuntimeOrigin: AsSystemOriginSigner<T::AccountId>,
+	{
+		const IDENTIFIER: &'static str = "ValidateUnsignedPriceSubmission";
+		type Implicit = ();
+		type Pre = ();
+		type Val = ();
+
+		fn weight(&self, _call: &Call<T>) -> frame_support::weights::Weight {
+			// Minimal weight since validation is lightweight
+			frame_support::weights::Weight::from_parts(5_000, 0)
+		}
+
+		fn validate(
+			&self,
+			origin: <Call<T> as Dispatchable>::RuntimeOrigin,
+			call: &Call<T>,
+			_info: &DispatchInfoOf<Call<T>>,
+			_len: usize,
+			_self_implicit: Self::Implicit,
+			_inherited_implication: &impl codec::Encode,
+			_source: TransactionSource,
+		) -> ValidateResult<Self::Val, Call<T>> {
+			// Check if this is an unsigned transaction (origin is None)
+			if origin.as_system_origin_signer().is_some() {
+				// This is a signed transaction, skip validation in this extension
+				return Ok((Default::default(), (), origin))
+			}
+
+			// This is an unsigned transaction - validate it
 			// Firstly let's check that we call the right function.
-			if let Call::submit_price_unsigned_with_signed_payload {
+			let validity = if let Call::submit_price_unsigned_with_signed_payload {
 				price_payload: ref payload,
 				ref signature,
 			} = call
 			{
 				let signature_valid =
 					SignedPayload::<T>::verify::<T::AuthorityId>(payload, signature.clone());
 				if !signature_valid {
-					return InvalidTransaction::BadProof.into()
+					return Err(InvalidTransaction::BadProof.into())
 				}
-				Self::validate_transaction_parameters(&payload.block_number, &payload.price)
+				Pallet::<T>::validate_transaction_parameters(&payload.block_number, &payload.price)
 			} else if let Call::submit_price_unsigned { block_number, price: new_price } = call {
-				Self::validate_transaction_parameters(block_number, new_price)
+				Pallet::<T>::validate_transaction_parameters(block_number, new_price)
 			} else {
-				InvalidTransaction::Call.into()
+				return Err(InvalidTransaction::Call.into())
```

Comment:
We can't return an error here, if the origin is not signed, it doesn't mean the call must be `submit_price_unsigned` or `submit_price_unsigned_with_signed_payload`.

There is 2 possible implementation:
* use `pallet::authorize_call`.
* match the call, if `submit_price_unsigned_with_signed_payload` or `submit_price_unsigned` then transmute the origin to some origin defined in this pallet (e.g. `ValidatedSubmitPrice`), in the call itself ensure the origin is signed or `ValidatedSubmitPrice`.